core/server/worker/src/main/java/alluxio/worker/block/UfsInputStreamManager.java

```diff
@@ -0,0 +1,300 @@
+                  try {
+                    inputStream.close();
+                  } catch (IOException e) {
+                    LOG.warn("Failed to close the input stream resource of file {} and resource id",
+                        inputStream.getFilePath(), removal.getKey());
+                  }
+                }
+                if (resources.isEmpty()) {
+                  // remove the resources entry
+                  mFileToInputStreamIds.remove(inputStream.getFilePath());
+                }
+              }
+            } else {
+              LOG.warn("Try to remove the resource entry of {} but not exists any more",
+                  removal.getKey());
+            }
+          }
+        };
+    mUnderFileInputStreamCache = CacheBuilder.newBuilder()
+        .maximumSize(Configuration.getInt(PropertyKey.WORKER_UFS_INSTREAM_CACHE_MAX_SIZE))
+        .expireAfterAccess(Configuration.getMs(PropertyKey.WORKER_UFS_INSTREAM_CACHE_EXPIRE_MS),
+            TimeUnit.MILLISECONDS)
+        .removalListener(RemovalListeners.asynchronous(listener, mRemovalThreadPool)).build();
+  }
+
+  /**
+   * Releases an input stream. The input stream is closed if it's already expired.
+   *
+   * @param inputStream the input stream to release
+   * @throws IOException when input stream fails to close
+   */
+  public void release(InputStream inputStream) throws IOException {
+    // for non-seekable input stream, close and return
+    if (!(inputStream instanceof SeekableUnderFileInputStream) || !isCachingEnabled()) {
+      inputStream.close();
+      return;
+    }
+
+    SeekableUnderFileInputStream seekableInputStream = (SeekableUnderFileInputStream) inputStream;
+    synchronized (mFileToInputStreamIds) {
+      if (!mFileToInputStreamIds.containsKey(seekableInputStream.getFilePath())) {
+        LOG.debug("The resource {} is already expired", seekableInputStream.getResourceId());
+        // the cache no longer tracks this input stream
+        seekableInputStream.close();
+        return;
+      }
+      UfsInputStreamIds resources = mFileToInputStreamIds.get(seekableInputStream.getFilePath());
+      if (!resources.release(seekableInputStream.getResourceId())) {
+        LOG.debug("Close the expired input stream resource of {}",
+            seekableInputStream.getResourceId());
+        // the input stream expired, close it
+        seekableInputStream.close();
+      }
+    }
+  }
+
+  /**
+   * Acquires an input stream. For seekable input streams, if there is an available input stream in
+   * the cache, reuse it and repositions the offset, otherwise the manager opens a new input stream.
+   *
+   * @param ufs the under file system
+   * @param path the path to the under storage file
+   * @param offset the offset to open
+   * @return the acquired input stream
+   * @throws IOException if the input stream fails to open
+   */
+  public InputStream acquire(UnderFileSystem ufs, String path, long offset) throws IOException {
```

Comment:
Should we pass `OpenOptions` instead of offset?

Reply:
Yup